shard-operator/src/main/java/com/redhat/service/bridge/shard/operator/BridgeIngressServiceImpl.java

```diff
@@ -80,26 +84,31 @@ public Deployment fetchOrCreateBridgeIngressDeployment(BridgeIngress bridgeIngre
         environmentVariables.add(new EnvVarBuilder().withName(KafkaConfigurationCostants.KAFKA_CLIENT_SECRET_ENV_VAR).withValue(kafkaConfigurationProvider.getSecret()).build());
         environmentVariables.add(new EnvVarBuilder().withName(KafkaConfigurationCostants.KAFKA_SECURITY_PROTOCOL_ENV_VAR).withValue(kafkaConfigurationProvider.getSecurityProtocol()).build());
         environmentVariables.add(new EnvVarBuilder().withName(Constants.BRIDGE_INGRESS_BRIDGE_ID_CONFIG_ENV_VAR).withValue(bridgeIngress.getSpec().getId()).build());
-        deployment.getSpec().getTemplate().getSpec().getContainers().get(0).setEnv(environmentVariables);
+        expected.getSpec().getTemplate().getSpec().getContainers().get(0).setEnv(environmentVariables);
+
+        Deployment existing = kubernetesClient.apps().deployments().inNamespace(bridgeIngress.getMetadata().getNamespace()).withName(bridgeIngress.getMetadata().getName()).get();
 
-        return kubernetesClient.apps().deployments().inNamespace(bridgeIngress.getMetadata().getNamespace()).create(deployment);
+        if (existing == null || !DeploymentSpecUtils.isDeploymentEqual(expected, existing)) {
+            return kubernetesClient.apps().deployments().inNamespace(bridgeIngress.getMetadata().getNamespace()).createOrReplace(expected);
+        }
+
+        return existing;
```

Comment:
The code in both services seems pretty much the same for me. Maybe we could refactor to a Kubernetes Resource Builder pattern? Instead of having `BridgeIngressService` and `BridgeExecutorService`, we would have `ShardServiceBuilder`, `ShardDeploymentBuilder`, and so on? wdyt?

Reply:
Hi @ricardozanini , the BridgeIngress and BridgeExecutor deployments are different. Some specific labels are set in the template itself (for example the `component` property that we use to filter the resources in the controllers), and some others are set by the `BridgeIngressService`/`BridgeExecutorService`, for example the environment variables are different. 
The `ShardServiceBuilder` and the `ShardDeploymentBuilder` can't really be generic, since the labels and some properties like the env variables need to be specific for the executor and the ingress, so I'm not sure it really improves the code (since we need to set them somewhere, probably again in a kind of `BridgeIngressService`). 
If it's fine for you I suggest we move forward with this PR and better discuss if we can improve the code in another ticket, is that ok for you @ricardozanini ?

Reply:
That's fine, this comment is not a blocker. We can discuss this further in another thread. But labels and env vars can be injected or used as a source to this new service. They are also pretty generic (literally a map). I think we should consider having the k8s resource creation centralized somehow. They both will share labels, envs, probes, and so on.